mozilla/thimble.mozilla.org
# Add webpack and convert homepage scripts to CommonJS

This patch is the first step in adding webpack to Thimble and switching away from requirejs and grunt.

### Changes in this patch
1. **Gruntfile.js** - Removes all the homepage scripts from the requirejs build step
2. **package.json** - Adds webpack and the npm scripts to run them
3. **public/homepage/scripts/(gallery | features | getinvolved).js** - None of the code has changed here. The only change made is to convert the require calls to CommonJS. It appears as a huge change due to indentation changes.
4. **public/homepage/scripts/main.js** - Requirejs config removed. Everything converted to commonjs. Removed passing around the modules via functions since they are required in.
5. **public/lib/(keyhandler | popupmenu | underlay).js** - These are identical copies to the files in `public/editor/scripts/editor/js/fc/` corresponding to `bramble-keyhandler.js`, `bramble-popupmenu.js` and `bramble-underlay.js`. I had to create a copy of these files and put them into `lib` because the code is shared with the scripts in `public/editor/scripts` which uses requirejs. Thus, to convert them into commonjs, I had to make copies of them. This is work that needs to happen anyway since it is shared code between the editor and the homepage.
6. **scripts/localize-client-webpack.js** - A temporary file that will exist till the webpack transition is completed. It calls functions from the regular `localize-client.js` file, however, we ignore the homepage scripts in `localize-client` since the source for localization is from `dist` (which is where the webpacked content will live) whereas `localize-client` enforces a source of `public`. Once we switch completely to webpack, this will disappear and the `localize-client` script will effectively turn into a webpack loader.

### Testing
1. Run `npm install` followed by `npm start`. Open up `localhost:3500` and try to navigate the website. On the homepage, check the network tab and make sure that only a single main.js file is loaded.
2. Change the entry for `NODE_ENV` in `.env` to `production`. Follow 1. above but make sure that the main.js file is minified.

### Notes
1. Watching the client for the homepage scripts no longer works. This is temporary and will be fixed once webpack is fully implemented as we will switch to webpack watch.
2. We will need to be careful about making changes to `public/editor/scripts/editor/js/fc/(bramble-keyhandler | bramble-popupmenu | bramble-underlay).js` to make sure that the same changes are made in their corresponding `public/lib` files.
